### PR TITLE
Add lempel ziv compression

### PR DESCRIPTION
**Pull request by iko1133**
* Implemented lempel-ziv-welch  compression/decompression algorithm

### CODE REVIEW

**Comment by iko1133** [comment]
The earlier pushed file ("decimal_to_any.py") seems to have broken the build...

8.64s$ flake8 --ignore=E203,W503 --max-complexity=25 --max-line-length=120 --statistics --count .

./conversions/decimal_to_any.py:70:25: E127 continuation line over-indented for visual indent

     E127 continuation line over-indented for visual indent

File: compression/lempel_ziv.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+"""
+    One of the several implementations of Lempel–Ziv–Welch compression algorithm
+    https://en.wikipedia.org/wiki/Lempel%E2%80%93Ziv%E2%80%93Welch
+"""
+
+import os
+import sys
+import math
```
Comment (cclauss) [inline compression/lempel_ziv.py:8]:
```suggestion
import math
import os
import sys
```
Sorted imports

File: compression/lempel_ziv.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def read_file_binary(file):
```
Comment (cclauss) [inline compression/lempel_ziv.py:11]:
```suggestion
def read_file_binary(file_path: str) -> str:
```
Add type hints as discussed in CONTRIBUTING.md.
Variable name should make it clear to the caller if we are looking for a file name or a file handle.

Are we returning bytes or str?!?.  Remember in Python 2 `bytes == str` but in Python 3 `bytes != str`

File: compression/lempel_ziv.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    Reads given file as bytes and returns them as a long string
+    """
+    result = ""
+    try:
+        with open(file, "rb") as binary_file:
+            data = binary_file.read()
+            for dat in data:
+                curr_byte = "{0:08b}".format(dat)
+                result += curr_byte
```
Comment (cclauss) [inline compression/lempel_ziv.py:21]:
```suggestion
        for dat in data:
            curr_byte = "{0:08b}".format(dat)
            result += curr_byte
```

File: compression/lempel_ziv.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return result
+    except IOError:
+        print("File not accessible")
+        sys.exit()
+
+
+def add_key_to_lexicon(lexicon, curr_string, index, last_match_id):
```
Comment (cclauss) [inline compression/lempel_ziv.py:28]: Type hints?  Doctests?  See CONTRIBUTING.md.

File: compression/lempel_ziv.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    lexicon[curr_string + "0"] = last_match_id
+
+    if math.log2(index).is_integer():
+        for curr_key in lexicon.keys():
+            lexicon[curr_key] = "0" + lexicon[curr_key]
+
+    lexicon[curr_string + "1"] = bin(index)[2:]
+
+
+def compress_data(data_bits):
```
Comment (cclauss) [inline compression/lempel_ziv.py:42]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        if curr_string not in lexicon.keys():
+            continue
+
+        last_match_id = lexicon[curr_string]
+        result += last_match_id
+        add_key_to_lexicon(lexicon, curr_string, index, last_match_id)
+        index += 1
+        curr_string = ""
+
+    while curr_string != "" and curr_string not in lexicon.keys():
```
Comment (cclauss) [inline compression/lempel_ziv.py:62]:
```suggestion
    while curr_string != "" and curr_string not in lexicon:
```
`dict.keys()` is almost never needed.

File: compression/lempel_ziv.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        curr_string += "0"
+
+    if curr_string != "":
+        last_match_id = lexicon[curr_string]
+        result += last_match_id
+
+    return result
+
+
+def add_file_length(source, compressed):
```
Comment (cclauss) [inline compression/lempel_ziv.py:72]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    Adds given file's length in front (using Elias  gamma coding) of the compressed string
+    """
+    file_length = os.path.getsize(source)
+    file_length_binary = bin(file_length)[2:]
+    length_length = len(file_length_binary)
+
+    return "0" * (length_length - 1) + file_length_binary + compressed
+
+
+def write_file_binary(file, to_write):
```
Comment (cclauss) [inline compression/lempel_ziv.py:83]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv_decompress.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
+"""
+    One of the several implementations of Lempel–Ziv–Welch decompression algorithm
+    https://en.wikipedia.org/wiki/Lempel%E2%80%93Ziv%E2%80%93Welch
+"""
+
+import sys
+import math
+
+
+def read_file_binary(file):
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:10]:
```suggestion
def read_file_binary(file_path: str) -> bytes:
```
Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv_decompress.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
+"""
+    One of the several implementations of Lempel–Ziv–Welch decompression algorithm
+    https://en.wikipedia.org/wiki/Lempel%E2%80%93Ziv%E2%80%93Welch
+"""
+
+import sys
+import math
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:7]:
```suggestion
import math
import sys
```

File: compression/lempel_ziv_decompress.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            for dat in data:
+                curr_byte = "{0:08b}".format(dat)
+                result += curr_byte
+        return result
+    except IOError:
+        print("File not accessible")
+        sys.exit()
+
+
+def decompress_data(data_bits):
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:27]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv_decompress.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                newLex["0" + curr_key] = lexicon.pop(curr_key)
+            lexicon = newLex
+
+        lexicon[bin(index)[2:]] = last_match_id + "1"
+        index += 1
+        curr_string = ""
+    return result
+
+
+def write_file_binary(file, to_write):
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:57]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv_decompress.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        if letter == "1":
+            break
+        counter += 1
+
+    data_bits = data_bits[counter:]
+    data_bits = data_bits[counter + 1 :]
+    return data_bits
+
+
+def compress(source, destination):
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:99]: Type hints? Doctests? See CONTRIBUTING.md.

File: compression/lempel_ziv.py (hunk 4 above)
Comment (iko1133) [reply compression/lempel_ziv.py:28]: Got it! sorry for that, sure I'll add those but I have a question: is Doctest still needed for a method/function, that does not return a value? and if so, can you give me an example? (this function is one of the ones, that don't return anything) 
(def add_key_to_lexicon(lexicon: dict, curr_string: str, index: int, last_match_id: int) -> None:
this is the new header for it)

Comment (cclauss) [reply compression/lempel_ziv.py]:
```
>>> my_key in lexicon
False
>>> add_key_to_lexicon(lexicon, my_key)
>>> my_key in lexicon
True
```

**Comment by cclauss** [comment]
Don't worry about `decimal_to_any.py`.  We will clean that up in post-production.

File: compression/lempel_ziv.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    """
+    Compresses given data_bits using Lempel–Ziv–Welch compression algorithm
+    and returns the result as a string
+    """
+    lexicon = {"0": "0", "1": "1"}
+    result, curr_string = "", ""
+    index = len(lexicon)
+
+    for i in range(len(data_bits)):
+        curr_string += data_bits[i]
```
Comment (cclauss) [inline compression/lempel_ziv.py:54]:
```suggestion
    for data_bit in data_bits:
        curr_string += data_bit
```

File: compression/lempel_ziv_decompress.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+        last_match_id = lexicon[curr_string]
+        result += last_match_id
+        lexicon[curr_string] = last_match_id + "0"
+
+        if math.log2(index).is_integer():
+            newLex = {}
+            for curr_key in list(lexicon):
+                newLex["0" + curr_key] = lexicon.pop(curr_key)
+            lexicon = newLex
```
Comment (cclauss) [inline compression/lempel_ziv_decompress.py:49]:
```suggestion
            lexicon = {"0" + key: key for key in lexicon}
```
Dict comprehension...

**Pull request merged**